lib/src/main/kotlin/StatefulSolver.kt

```diff
@@ -1,105 +1,82 @@
-import kotlin.math.ln
 import kotlin.math.max
-import kotlin.math.sqrt
-import kotlin.random.Random
-import kotlin.text.StringBuilder
-
-class StatefulSolver<TState, TAction>(
-        private val mdp: MDP<TState, TAction>,
-        private val random: Random,
-        private val iterations: Int,
-        private val simulationDepthLimit: Int,
-        private val explorationConstant: Double,
-        private val rewardDiscountFactor: Double,
-        private val verbose: Boolean) {
-
-    private var root : StateNode<TAction, TState>? = null
-
-    fun buildTree() {
-        initialize()
-
-        for (i in 0 until iterations) {
-            iterateStep()
-        }
-    }
-
-    fun initialize() {
-        val initialState = mdp.initialState()
-        root = createStateNode(null, initialState)
-    }
 
-    fun iterateStep() {
-        traceln("")
-        traceln("New iteration")
-        traceln("=============")
+open class StatefulSolver<TState, TAction>(
+    private val mdp: MDP<TState, TAction>,
+    private val simulationDepthLimit: Int,
+    explorationConstant: Double,
+    private val rewardDiscountFactor: Double,
+    verbose: Boolean
+) : SolverBase<TAction, StateNode<TState, TAction>>(verbose, explorationConstant) {
 
-        // Selection
-        val bestState = selectNode(root!!)
+    override var _root = createNode(null, null, mdp.initialState())
 
-        if (verbose) {
-            traceln("Selected:")
-            displayNode(bestState)
+    override fun selectNode(node: StateNode<TState, TAction>): StateNode<TState, TAction> {
+        // If the node is terminal, return it
+        if (mdp.isTerminal(node.state)) {
+            return node
         }
 
-        // Expansion
-        val expandedState = expandNode(bestState)
+        val exploredActions = node.exploredActions()
 
-        if (verbose) {
-            traceln("Expanding:")
-            displayNode(expandedState)
+        // This state has not been fully explored
+        if (node.validActions.size != exploredActions.size) {
```

Comment:
Safer not? `if (node.validActions.size > exploredActions.size)`

Reply:
Yea probably. though valid actions can never be less than explored actions. I might add a debug assert here or throw an exception when that occurs.